app/forms/payment_form.rb

```diff
@@ -0,0 +1,82 @@
+class PaymentForm < ApplicationForm
+  attr_accessor :account_transaction,
+                :processor_token,
+                :coupon,
+                :user
+
+  before_save :decrease_stock
+  before_save :mark_coupon_as_used
+  before_save :charge!
```

Comment:
I think this things should happen `after_save`, otherwise we could be left on a scenario that the charge was performed in the external service but our system don't know. I mean that instead of

```
1. charge on external service
2. save models
```
doing
```
1. save models
2. charge on external service
3. if error on external service update the model again
```